### PR TITLE
Feature/type hints

### PR DESCRIPTION
**Pull request by robert-richardson**
**General recommendations:**

- Use type hints - they are extremely useful: for others to understand your code quickly, for future you to understand your code quickly, for current you to have vastly more helpful hints and autocomplete features as you write the code within an IDE like pycharm
- Use “click” rather than sys.argv. It’s much more flexible, and allows for example to do automatic conversion  and checking of arguments passed to the script (e.g. a date string can be converted to a dt.date or pd.Timestamp object when passed to the function, and a path in the form of a string can be converted to a Path object directly)
- Setup mypy for type checking - it will prevent bugs and tech debt in the future
- Create a few (need not be several, just judicious choice) smaller tests, which do not require reading any external data
- Rename variables more verbose, e.g. g_hh -> gen_hh, l_hh -> load_hh. Many engineers would be skewed even more towards verbose, i.e. load_halfhourly, but I think it’s ok to keep it as hh. Also “d” is very ambiguous (especially in the absence of type hints). Does it imply dictionary or data frame or other!? Use df for dataframes.
- Use the pathlib module, more modern and concise than os.path, os.listdir etc.. Anywhere you use os.path, os.listdir etc., Google (or ask chatgpt) how to do it with pathlib.Path. Likewise anywhere you pass around a plain str object which is actually a representing a path, instead pass the Path object.
- try to avoid second order functions (i.e. passing functions as argument to another function) unless it’s really needed. It makes following the data flow much harder and quickly loses type hint information in pycharm. Also, in my personal experience at least, it’s rarely actually necessary
- Pass pd.Series rather than pd.Dataframe -> it’s much easier to follow the flow because it doesn’t contain an unknown number of columns and column names
- In some places you use nested dictionaries and tuples, which is very hard to follow. For these, it’s much better to use dataclasses or Attr classes, to explicitly define the structure of the objects

### CODE REVIEW

File: scores/s0142/tests/test_s0142.py [hunk 1]
```diff
@@ -11,15 +12,15 @@ def test_process_file():
...
+        / "raw/S0142/S0142_20220401_SF_20220427115520.gz",
+        bsc_party_ids=list(expected.keys()),
     ):
         result = load.sum()
         assert result["Settlement Period"] == 1176  # 1 + 2 + 3 + ... + 48
         for metric_key, metric_value in expected[bsc_party_id].items():
             assert result[metric_key] == pytest.approx(metric_value, 0.01)
+
+
+if __name__ == "__main__":
```
Comment (robert-richardson) [inline scores/s0142/tests/test_s0142.py:25]: inadvertently leftover (todo: remove)

**Review by joejk2 — APPROVED** [review]
🙌

🙏
